File: solutions/0/0/7/3/p0073_s2.c
/* Solution to Euler Problem 73: Counting Fractions in a Range. */
#include "runner.h"

/* Rank of n/d in the Farey sequence F_max_d: the count of reduced fractions in [0, 1] with
   denominator at most max_d that do not exceed n/d. Builds data[i] = floor(i*n/d) (all fractions,
   reduced or not), then an additive Mobius sieve subtracts each data[i] from its multiples,
   leaving the exact coprime count per denominator. O(max_d log max_d). */
static long long rank_fn(int max_d, int n, int d) {
    int len_data = max_d + 1;
    long long *data = malloc((size_t)len_data * sizeof(long long));
    if (!data) return -1;

    for (int i = 0; i < len_data; i++) {
        /* Cast to long long before the product to avoid 32-bit overflow when i is large. */
        data[i] = (long long)i * n / d;
    }

    /* Additive sieve (Mobius inversion): strip multiples so data[i] counts only denominator i. */
    for (int i = 1; i < len_data; i++) {
        for (int j = 2 * i; j < len_data; j += i) {
            data[j] -= data[i];
        }
    }

    long long total = 0;
    for (int i = 0; i < len_data; i++) {
        total += data[i];
    }

    free(data);
    return total;
}

const char *solve(int argc, char *argv[]) {
    /* Counting fractions in (1/3, 1/2) is a rank difference: rank(1/2) - rank(1/3) - 1, where the
       -1 drops 1/2 itself. Each rank is an O(max_d log max_d) Mobius sieve. */
    static char _answer[32];
    int max_d = parse_int(argv[1]);
    { snprintf(_answer, sizeof _answer, "%lld", (long long)(rank_fn(max_d, 1, 2) - rank_fn(max_d, 1, 3) - 1)); return _answer; }
}

File: solutions/public/p0073/p0073_s2.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 73: Counting Fractions in a Range [Level 3]. """
from __future__ import annotations

from solver.runners import runner


@runner.main
def solve(*args: str) -> str:
    """Count fractions in (1/3, 1/2) as a Farey rank difference: rank(1/2) - rank(1/3) - 1 (the -1
    drops 1/2 itself), each rank computed by an O(max_d log max_d) additive Mobius sieve."""
    max_d = runner.parse_int(args[0])

    def rank(n: int, d: int) -> int:
        """Reduced fractions with denominator <= max_d that are <= n/d, via an additive Mobius
        sieve: data[i] = floor(i*n/d) counts all such fractions, then subtracting each data[i]
        from its multiples leaves the exact coprime count per denominator; their sum is the rank."""
        len_data: int = max_d + 1
        data: list[int] = [i * n // d for i in range(len_data)]
        for i in range(1, len_data):
            for j in range(2 * i, len_data, i):
                data[j] -= data[i]
        return sum(data)

    return str(rank(n=1, d=2) - rank(n=1, d=3) - 1)


if __name__ == "__main__":
    raise SystemExit(solve())
